allennlp/common/from_params.py

```diff
@@ -0,0 +1,185 @@
+
+        # This handles types like Dict[str, TokenIndexer], where we need to instantiate
+        # each value from_params and return the resulting dict.
+        elif origin == Dict and len(args) == 2 and hasattr(args[-1], 'from_params'):
+            value_cls = annotation.__args__[-1]
+
+            value_dict = {}
+
+            for key, value_params in params.pop(name, Params({})).items():
+                value_dict[key] = value_cls.from_params(params=value_params, **extras)
+
+            kwargs[name] = value_dict
+
+        else:
+            # Pass it on as is and hope for the best.   ¯\_(ツ)_/¯
+            if optional:
+                kwargs[name] = params.pop(name, default)
+            else:
+                kwargs[name] = params.pop(name)
+
+    params.assert_empty(cls.__name__)
+    return {k: v for k, v in kwargs.items() if takes_arg(cls, k)}
+
+
+
+class FromParams:
+    """
+    Mixin to give a from_params method to classes. We create a distinct base class for this
+    because sometimes we want non-Registrable classes to be instantiatable from_params.
+    """
+    @classmethod
+    def from_params(cls: Type[T], params: Params, **extras) -> T:
+        """
+        This is the automatic implementation of `from_params`. Any class that subclasses `FromParams`
+        (or `Registrable`, which itself subclasses `FromParams`) gets this implementation for free.
+        If you want your class to be instantiated from params in the "obvious" way -- pop off parameters
+        and hand them to your constructor with the same names -- this provides that functionality.
+
+        If you need more complex logic in your from `from_params` method, you'll have to implement
+        your own method that overrides this one.
+        """
+        # pylint: disable=protected-access
+        from allennlp.common.registrable import Registrable  # import here to avoid circular imports
+
+        logger.info(f"instantiating class {cls} from params {getattr(params, 'params', params)} "
+                    f"and extras {extras}")
+
+        if params is None:
+            return None
+
+        registered_subclasses = Registrable._registry.get(cls)
+
+        if registered_subclasses is not None:
+            # We know ``cls`` inherits from Registrable, so we'll use a cast to make mypy happy.
+            # We have to use a disable to make pylint happy.
+            # pylint: disable=no-member
+            as_registrable = cast(Type[Registrable], cls)
+            default_to_first_choice = as_registrable.default_implementation is not None
+            choice = params.pop_choice("type",
+                                       choices=as_registrable.list_available(),
+                                       default_to_first_choice=default_to_first_choice)
+            subclass = registered_subclasses[choice]
+
+            # Prune down the dict of extras, if necessary.
+            if not takes_arg(subclass.from_params, 'extras'):
```

Comment:
Again, an example of when you need this here would be nice.  Do we have `Registrable` classes where some subclasses take extra args that others don't?  That seems like it breaks the old `from_params` API, so I'm not sure where it ever happened.

Reply:
it's a little bit more subtle than that, I added some explanation